ADAL/src/cache/ADMSIDDataSourceWrapper.m

```diff
@@ -217,6 +219,66 @@ - (ADTokenCacheItem *)getItemWithKey:(ADTokenCacheKey *)key
     return results;
 }
 
+- (BOOL)removeAllForClientId:(NSString *)clientId
+                       error:(ADAuthenticationError **)error
+{
+    MSID_LOG_WARN(nil, @"Removing all items for client");
+    MSID_LOG_WARN_PII(nil, @"Removing all items for client %@", clientId);
+    
+    return [self removeAllForUserIdImpl:nil clientId:clientId error:error];
+}
+
+
+- (BOOL)removeAllForUserId:(NSString *)userId
+                  clientId:(NSString *)clientId
+                     error:(ADAuthenticationError **)error
+{
+    MSID_LOG_WARN(nil, @"Removing all items for user");
+    MSID_LOG_WARN_PII(nil, @"Removing all items for user + client <%@> userid <%@>", clientId, userId);
+    
+    return [self removeAllForUserIdImpl:userId clientId:clientId error:error];
+}
+
+- (BOOL)wipeAllItemsForUserId:(NSString *)userId
+                        error:(ADAuthenticationError **)error
+{
+    MSID_LOG_WARN(nil, @"Removing all items for user.");
+    MSID_LOG_WARN_PII(nil, @"Removing all items for userId <%@>", userId);
+    
+    BOOL result = [self removeAllForUserIdImpl:userId clientId:nil error:error];
```

Comment:
Can we use previous wipe logic of composing a query and removing items for a query? This would optimize it for keychain from 1 read and 1 delete operation to 1 delete operation.

Reply:
That optimization logic can only work for iOS because `[MSIDMacTokenCache removeItemsWithKey:...]` do not support nil/wild card authority/clientid/resource, as `MSIDMacTokenCache ` uses `MSIDLegacyTokenCacheKey` object as the dictionary key.

Should we change `[MSIDMacTokenCache removeItemsWithKey:...]` or should we have a separate wipe logic for keychain token cache? :)

Reply:
We can leave it the way it is for now